sdk/src/main/scala/com/horizen/SidechainState.scala

```diff
@@ -266,14 +266,28 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage,
         }
       }
 
-      newCoinsBoxesAmount = tx.newBoxes().asScala
+      val newBoxes = tx.newBoxes().asScala
+
+      newCoinsBoxesAmount = newBoxes
         .filter(box => box.isInstanceOf[CoinsBox[_ <: PublicKey25519Proposition]] || box.isInstanceOf[WithdrawalRequestBox])
         .map(_.value()).sum
 
       if (closedCoinsBoxesAmount != newCoinsBoxesAmount + tx.fee())
         throw new Exception("Amounts sum of CoinsBoxes is incorrect. " +
           s"ClosedBox amount - $closedCoinsBoxesAmount, NewBoxesAmount - $newCoinsBoxesAmount, Fee - ${tx.fee()}")
 
+      newBoxes
+        .filter(box => box.isInstanceOf[ForgerBox])
+        .foreach(forgerBox => {
+          if (params.restrictForgers) {
+            val vrfPublicKey: VrfPublicKey = forgerBox.vrfPubKey()
+            val blockSignProposition: PublicKey25519Proposition = forgerBox.blockSignProposition()
+            if (!params.allowedForgersList.contains((blockSignProposition, vrfPublicKey))) {
```

Comment:
Why restricting both vrfPubKey and blockSignProposition?

Reply:
Forger stake identified by the pair of keys: vrf and block sign.
If we will check only against the one of them, for example, block sign, than still no one except the trusted forger will be able to create a new block, because of the lack of private key for the other public key.
But it will allow malicious actor to create the new forger stakes, that will decrease the blocks generation rate.
Possible attack: honest forgers have a total stake of 10 zen -> someone create a fake stake with 100 zen -> forging rate will be decreased almost 10 times.